modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js

```diff
@@ -0,0 +1,211 @@
+
+    switch (this.type) {
+      case 1: // Point
+        this._firstPassData.pointFeaturesCount++;
+        this._firstPassData.pointPositionsCount += geom.lines.length;
+        break;
+
+      case 2: // LineString
+        this._firstPassData.lineFeaturesCount++;
+        this._firstPassData.linePathsCount += geom.lines.length;
+        this._firstPassData.linePositionsCount += geom.data.length / coordLength;
+        break;
+
+      case 3: // Polygon
+        const rings = classifyRings(geom.lines);
+        this._firstPassData.polygonFeaturesCount++;
+        this._firstPassData.polygonObjectsCount += geom.lines.length;
+
+        for (const lines of rings) {
+          this._firstPassData.polygonRingsCount += lines.length;
+        }
+        this._firstPassData.polygonPositionsCount += geom.data.length / coordLength;
+
+        // Unlike Point & LineString geom.lines is a 2D array, thanks
+        // to the classifyRings method
+        geom.lines = rings;
+        break;
+    }
+
+    geom.type = VectorTileFeature.types[this.type];
+    if (geom.lines.length > 1) {
+      geom.type = `Multi${geom.type}`;
+    }
+
+    const result = {
+      type: 'Feature',
+      geometry: geom,
+      properties: this.properties
+    };
+
+    if (this.id !== null) {
+      result.id = this.id;
+    }
+
+    return result;
+  }
+
+  toLines(options) {
+    if (typeof options === 'function') {
+      return this._toLines(options);
+    }
+    const {x, y, z} = options;
+    const size = this.extent * Math.pow(2, z);
+    const x0 = this.extent * x;
+    const y0 = this.extent * y;
+
+    function project(data) {
+      for (let j = 0, jl = data.length; j < jl; j += 2) {
+        data[j] = ((data[j] + x0) * 360) / size - 180;
+        const y2 = 180 - ((data[j + 1] + y0) * 360) / size;
+        data[j + 1] = (360 / Math.PI) * Math.atan(Math.exp((y2 * Math.PI) / 180)) - 90;
+      }
+    }
+    return this._toLines(project);
+  }
+}
+
+/**
+ * In the original geojson implementation this function
+ * classifies the rings in a polygon into outlines & holes,
+ * based on the vertex winding order. However the implementation
+ * was broken and all it did was classify each ring as a separate
+ * polygon outline. In order to maintain compatibility with the
+ * original, we keep this behavior, but simplify the implementation
+ * to an equivalent function:
+ * [A, B, C] => [[A], [B], [C]]
+ *
+ * TODO in the future, the classification could be fixed following
+ * https://github.com/mapbox/vector-tile-js, but this would require
+ * confirming that the renderer of deck.gl is actually capable of
+ * drawing polygons with holes in them.
+ */
+function classifyRings(rings) {
+  return rings.map(r => [r]);
+}
+
+function readFeature(tag, feature, pbf) {
+  if (tag === 1) feature.id = pbf.readVarint();
+  else if (tag === 2) readTag(pbf, feature);
+  else if (tag === 3) feature.type = pbf.readVarint();
+  else if (tag === 4) feature._geometry = pbf.pos;
```

Comment:
Nit: use a switch here?

Reply:
This, like a lot of the code is just what was in the original mapbox implementation, I think it is worth keeping it the same, as it makes it easier to reference against the original code